elastic/beats
# Add numCPU wrapper for more accurate metrics reporting

## What does this PR do?

This deals with a long-standing issue in how we collect metrics. `runtime.NumCPU()` isn't a sound API for gathering actual CPU information, as on most platforms it uses CPU affinity APIs to derive the CPU count. For most use cases (figuring out how many threads we should start, etc), this is fine, but if we want actual _metrics_ on CPU counts, this isn't great, as VMs, containerization, certain job control systems, user tinkering, etc, can mess with this value. This adds a new `numcpu` library that works as a drop-in replacement for `runtime.NumCPU()` It has a few OS-specific implementations, and it'll fallback to `NumCPU` when it needs to.

I didn't quite know enough about Windows to know how to properly implement it, so I just used the same API we use elsewhere for gathering per-CPU metrics, and counted the CPUs from there. @narph I assume there's a better way to do this?

## Why is it important?
This has been an issue in the past with metricbeat reporting incorrect metrics in the rare cases where CPU affinity does not reflect hardware. This fixes that.

## Checklist

## How to test this PR locally

- For now, just test to make sure metrics like `system.load.cores` report the correct values. I'm gonna see if I can come up with a more sophisticated test for CPU affinity...
